File: Java/Migration/unit-tests-csharp/unit-tests-csharp/P04NSubstitute/Ex01BasicStubbingAndMocking.cs
using NSubstitute;
using NUnit.Framework;
using TddEbook.TddToolkit;
using unit_tests_csharp.P04NSubstitute.ProductionCode;

namespace unit_tests_csharp.P04NSubstitute
{
  public class Ex01BasicStubbingAndMocking
  {
    [Test]
    public void ShouldCopyDataFromSourceToDestination()
    {
      //GIVEN
      var copyOperation = new CopyOperation();
      var destination = Substitute.For<IDataDestination>();
      var source = Substitute.For<IDataSource>();
      var data = Any.Instance<Data>();

      source.RetrieveData().Returns(data);

      //WHEN
      copyOperation.ApplyTo(source, destination);

      //THEN
      destination.Received(1).Save(data);
      //destination.Received().Save(data);
    }
  }
}

File: Java/xunit-test-patterns/src/test/java/_09_mock_objects/Ex01BasicStubbingAndMocking.java
package _09_mock_objects;

import autofixture.publicinterface.Any;
import lombok.val;
import org.junit.jupiter.api.Test;

import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class Ex01BasicStubbingAndMocking {
    @Test
    public void shouldCopyDataFromSourceToDestination() {
        //GIVEN
        val copyOperation = new CopyOperation();
        val destination = mock(DataDestination.class);
        val source = mock(DataSource.class);
        val data = Any.anonymous(Data.class);

        when(source.retrieveData()).thenReturn(data);

        //WHEN
        copyOperation.applyTo(source, destination);

        //THEN
        verify(destination).save(data);
        //verify(destination /*, atLeast(1)*/).save(data);
    }

    @Test
    public void shouldCopyDataFromSourceToDestination_bdd() {
        //GIVEN
        val copyOperation = new CopyOperation();
        val destination = mock(DataDestination.class);
        val source = mock(DataSource.class);
        val data = Any.anonymous(Data.class);

        given(source.retrieveData()).willReturn(data);

        //WHEN
        copyOperation.applyTo(source, destination);

        //THEN
        then(destination).should().save(data);
    }
}
